tools/dumppdf.py

```diff
@@ -128,51 +129,186 @@ def dumpallobjs(out, doc, mode=None):
     out.write('</pdf>')
     return
 
-# dumpoutline
+
+def resolve_dest(dest, doc):
+    if isinstance(dest, str):
+        dest = resolve1(doc.get_dest(dest))
+    elif isinstance(dest, PSLiteral):
+        dest = resolve1(doc.get_dest(dest.name))
+    if isinstance(dest, dict):
+        dest = dest['D']
+    return dest
+
+
+def get_page_number(dest, doc, pages, a):
+    pageno = None
+    if dest:
+        dest = resolve_dest(dest, doc)
+        pageno = pages[dest[0].objid]
+    elif a:
+        action = a.resolve()
+        if isinstance(action, dict):
+            subtype = action.get('S')
+            if subtype and repr(
+                    subtype) == '/GoTo' and action.get('D'):
+                dest = resolve_dest(action['D'], doc)
+                pageno = pages[dest[0].objid]
+    return pageno
+
+
+def extract_outline_info(pages, outlines, doc, outfp, dest_info):
```

Comment:
Would be nice to know what the arguments are expected to be, as well as what the method does